### ISSUE TITLE
Using refreshRate with non-existent labels causes an existent branch to be staled

### ISSUE THREAD

**Issue by josebarrueta-okta** [spring-cloud/spring-cloud-config#1174]
We found this problem using in the spring-cloud-config-sever v1.4.4.RELEASE.

We have a Cloud Config client that is configured to use fallback labels when talking to the CCS, for example: `spring.cloud.config.label=non-existing-branch,test-staled-branch`

When the Cloud Config server is configured to use a `refreshRate` for example `10s`, we noticed that the `git` updates in the `test-staled-branch` are never pulled in the server and therefore not sent to the client.

This behaviour is reproducible all the time and when we were debugging we found that the issue is that the when the client request the `non-existing-branch` label, that call updates the `lastRefresh` timestamp, since this call returns `404` to the client, the next label is attempted in this example: `test-staled-branch`, but since the previous call already updated the `lastRefresh` timestamp, the check for `shouldPull` returns `false` causing the existent branch to be staled.

```
	protected boolean shouldPull(Git git) throws GitAPIException {
		boolean shouldPull;

		if (this.refreshRate > 0 && System.currentTimeMillis() - this.lastRefresh < (this.refreshRate * 1000)) {
			return false;
		}
        ...
```

My first approach to solve this issue is to create a `refreshRate` by label instead of by repository

```
@ConfigurationProperties("spring.cloud.config.server.git")
public class LabeledJGitEnvironmentRepository extends MultipleJGitEnvironmentRepository {

    private static final ThreadLocal<String> GIT_LABEL_HOLDER = new ThreadLocal<>();
    
    private Map<String, Long> labelRefreshMap;

    public LabeledJGitEnvironmentRepository(ConfigurableEnvironment environment) {
        super(environment);
        this.labelRefreshMap = new ConcurrentHashMap<>();
    }

    @Override
    public String refresh(String label) {
        try {
            GIT_LABEL_HOLDER.set(label);
            return super.refresh(label);
        } finally {
            GIT_LABEL_HOLDER.remove();
        }
    }

    @Override
    public long getLastRefresh() {
        return labelRefreshMap.getOrDefault(GIT_LABEL_HOLDER.get(), 0L);
    }

    @Override
    public void setLastRefresh(long lastRefresh) {
        labelRefreshMap.put(GIT_LABEL_HOLDER.get(), lastRefresh);
    }

    @Override
    protected boolean shouldPull(Git git) throws GitAPIException {
        boolean shouldPull;

        if (this.getRefreshRate() > 0 && System.currentTimeMillis() - this.getLastRefresh() < (this.getRefreshRate() * 1000)) {
            return false;
        }
     ...
    }
```

Which works but might result in more fetches than necessary, still was easy to make it work. 

While we are here, would be nice to take the `refreshRate` logic to its own `protected` method so it's easier to override the logic, in our use case we wanted to have an easy way to keep the refreshes synchronised across multiple CCS in a cluster using the `System.currentTimeMillis()`

Thanks!

**Comment by ryanjbaxter** [comment]
Maybe the refresh rate should only be updated if the fetch call succeeds.

**Comment by josebarrueta-okta** [comment]
That's another approach that will work too, it will have the same effect as my solution, with fetch being executed multiple times until the first existent label is requested.

This is the current code:

```
	public String refresh(String label) {
		Git git = null;
		try {
			git = createGitClient();
			if (shouldPull(git)) {
				FetchResult fetchStatus = fetch(git, label);
				if (deleteUntrackedBranches && fetchStatus != null) {
					deleteUntrackedLocalBranches(fetchStatus.getTrackingRefUpdates(), git);
				}
				// checkout after fetch so we can get any new branches, tags, ect.
				checkout(git, label);
				if (isBranch(git, label)) {
					// merge results from fetch
					merge(git, label);
					if (!isClean(git)) {
						logger.warn(
								"The local repository is dirty. Resetting it to origin/"
										+ label + ".");
						resetHard(git, label, LOCAL_BRANCH_REF_PREFIX + label);
					}
				}
			}
			else {
				// nothing to update so just checkout
				checkout(git, label);
			}
			// always return what is currently HEAD as the version
			return git.getRepository().getRef("HEAD").getObjectId().getName();
		}
``` 

I think the approach that works better is to decouple the `fetch` (git call) from the `merge` logic (applying the changes).

I think it's not as simple as I am putting it here, because `shouldPull` also checks for `dirty` branches etc, but if we take into account that the `refreshRate` is trying to avoid calls to the remote server, we can provide split the logic to something like:

```
    public String refresh(String label) {
        Git git = null;
        try {
            git = createGitClient();
            if (hasRefreshRateElapsed()) {
                FetchResult fetchStatus = fetch(git, label);
                if (deleteUntrackedBranches && fetchStatus != null) {
                    deleteUntrackedLocalBranches(fetchStatus.getTrackingRefUpdates(), git);
                }
            }
            
            if (shouldPull(git)) {
                // checkout after fetch so we can get any new branches, tags, ect.
                checkout(git, label);
                if (isBranch(git, label)) {
                    // merge results from fetch
                    merge(git, label);
                    if (!isClean(git)) {
                        logger.warn(
                                "The local repository is dirty. Resetting it to origin/"
                                        + label + ".");
                        resetHard(git, label, LOCAL_BRANCH_REF_PREFIX + label);
                    }
                }
            }
            else {
                // nothing to update so just checkout
                checkout(git, label);
            }
       ....
    }

   protected boolean shouldPull(Git git) throws GitAPIException {
        boolean shouldPull;

        //remove the refresh logic from here

        Status gitStatus = git.status().call();
        boolean isWorkingTreeClean = gitStatus.isClean();
        String originUrl = git.getRepository().getConfig().getString("remote", "origin", "url");

        if (this.isForcePull() && !isWorkingTreeClean) {
            shouldPull = true;
            logDirty(gitStatus);
        } else {
            shouldPull = isWorkingTreeClean && originUrl != null;
        }
        if (!isWorkingTreeClean && !this.isForcePull()) {
            this.logger.info("Cannot pull from remote " + originUrl
                    + ", the working tree is not clean.");
        }
        return shouldPull;
    }

   protected boolean hasRefreshRateElapsed() {
    ...
  }
```

Like I said probably I might be missing something but i really think something likes this works better.

**Comment by ryanjbaxter** [comment]
Want to submit a PR?

**Comment by josebarrueta-okta** [comment]
I'll be happy to submit a PR.

For our use case (and until we have the fix is release), our solution looks like this (extending `MultipleJGitEnvironmentRepository`

```
@Override
    protected boolean shouldPull(Git git) throws GitAPIException {
        boolean shouldPull;
        //remove refreshRate check from this method
        Status gitStatus = git.status().call();
        boolean isWorkingTreeClean = gitStatus.isClean();
        String originUrl = git.getRepository().getConfig().getString("remote", "origin", "url");

        if (this.isForcePull() && !isWorkingTreeClean) {
            shouldPull = true;
            logDirty(gitStatus);
        } else {
            shouldPull = isWorkingTreeClean && originUrl != null;
        }
        if (!isWorkingTreeClean && !this.isForcePull()) {
            this.logger.info("Cannot pull from remote " + originUrl + ", the working tree is not clean.");
        }
        return shouldPull;
    }

    @Override
    protected FetchResult fetch(Git git, String label) {
        if (!hasRefreshRateElapsed()) {
            return null;
        }
        return super.fetch(git, label);
    }

    protected boolean hasRefreshRateElapsed() {
        int refreshRate = this.getRefreshRate();
        if (refreshRate <= 0) {
            return true;
        }
        ...
    }
```

So it only fetches once, ever `refreshRate` seconds but it merges the changes in the existent label.

**Comment by josebarrueta-okta** [comment]
Hi @ryanjbaxter , when checkout the code realised that this was fixed in 1.4.5, so this can be closed:


https://github.com/spring-cloud/spring-cloud-config/commit/80f98c387af5b1a08228b4f19ac411f3d5b22ae5#diff-3c25894c5eacbf4d83e593f438b9469c

**Comment by izeye** [comment]
Based on the above comment from @josebarrueta-okta , it seems the milestone needs to be unassigned.
